### PR TITLE
Make the service worker check for itself and unregister if not present

### LINKED ISSUE #2: Make the service worker to check for itself to unregister when a 404 is returned.
If the user disable all the plugins, the service worker is not unregistered.

### PR DESCRIPTION
**Pull request by delapuente**
Fix #2

### CODE REVIEW

**Comment by delapuente** [comment]
@mykmelez do you want to take a look about how to make the sw unregister itself when there is no longer present in the server?

File: lib/js/sw-base.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          timestamp = timestamp || 0;
+          var now = Date.now();
+          if (now - timestamp > this.ONE_DAY) {
+            return Promise.all([
+              this.setLastCheck(now),
+              fetch(self.registration.active.scriptURL, { method: 'HEAD' })
+            ])
+            .then(results => {
+              var response = results[1];
+              // If not served as JavaScript, it is the error page from WordPress
```
Comment (delapuente) [inline lib/js/sw-base.js]: Will be possible to check for 404 when #7 lands.

File: lib/js/sw-base.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+              return response.headers.get('Content-Type').indexOf('javascript') < 0;
```
Comment (mykmelez) [inline lib/js/sw-base.js]: I would `toLowerCase()` before `indexOf`. But I'm not sure how reliable this check will be. If browsers are lenient about enforcing this, then I can imagine it being unreliable.

File: lib/js/sw-base.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      .then(() => this._lastCheck = Promise.resolve(value));
+    },
+
+    onFetch: function (event) {
+      event.waitUntil(this.checkForUnregistering().then(shouldUnregister => {
+        if (shouldUnregister) {
+          this.storage.clear();
+          return this.unregister();
+        }
+        return Promise.resolve();
```
Comment (mykmelez) [inline lib/js/sw-base.js:55]: I might cache a resolved promise to return here, for perf. Unsure if it matters, though.

**Comment by mykmelez** [comment]
Overall looks good. My only concern is around perf. I wonder about the cost of intercepting every fetch and waiting until the resolution of a promise. I don't necessarily have a better idea, but it'd be helpful to get a sense of how much time is spent in that part of the code, to ensure that it's not significant (or find ways to reduce the cost if it is significant).

**Comment by delapuente** [comment]
Although it is a source of concerns, it is palliated by the fact we are not delaying responses, just keeping the event alive a little bit more to finish that request. Critical sections here are the access to IDB and the `HEAD` request. First is cached and performed only once by worker execution which is reasonable and the other is performed only once in 24h. I though about caching all the results but then I opted for recalculating the expiration as I don't think it's adding a noticeable overhead and thus, to cover some corner cases.

File: lib/js/sw-base.js (hunk 3 above)
Comment (marco-c) [inline lib/js/sw-base.js:55]: You could also avoid returning a promise here.

Comment (delapuente) [inline lib/js/sw-base.js:55]: I prefer to be coherent with my return types.

**Pull request merged**